diff --git a/dlls/ntdll/unix/signal_x86_64.c b/dlls/ntdll/unix/signal_x86_64.c
index abb98f7..b6cd912 100644
--- a/dlls/ntdll/unix/signal_x86_64.c
+++ b/dlls/ntdll/unix/signal_x86_64.c
@@ -2029,6 +2029,7 @@ static void install_bpf(struct sigaction *sig_act)
     unsigned int i, j;
     NTSTATUS status;
 
+#ifndef __ANDROID__
     if ((ULONG_PTR)sc_seccomp < NATIVE_SYSCALL_ADDRESS_START
             || (ULONG_PTR)syscall < NATIVE_SYSCALL_ADDRESS_START)
     {
@@ -2037,6 +2038,7 @@ static void install_bpf(struct sigaction *sig_act)
         ERR_(seh)("The known reasons are /proc/sys/vm/legacy_va_layout set to 1 or 'ulimit -s' being 'unlimited'.\n");
         return;
     }
+#endif
 
     sig_act->sa_sigaction = sigsys_handler;
     memset(&prog, 0, sizeof(prog));
@@ -2067,9 +2069,15 @@ static void install_bpf(struct sigaction *sig_act)
 
     frame->syscall_flags = syscall_flags;
 
+#ifndef __ANDROID__
     test_syscall = mmap((void *)0x600000000000, 0x1000, PROT_EXEC | PROT_READ | PROT_WRITE,
             MAP_PRIVATE | MAP_ANON, -1, 0);
     if (test_syscall != (void *)0x600000000000)
+#else
+    test_syscall = mmap(NULL, 0x1000, PROT_EXEC | PROT_READ | PROT_WRITE,
+            MAP_PRIVATE | MAP_ANON, -1, 0);
+    if (test_syscall == MAP_FAILED)
+#endif
     {
         int ret;
 
