diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 28a9f5f..27104a9 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -569,7 +569,9 @@ static void sync_window_style( struct x11drv_win_data *data )
         int mask = get_window_attributes( data, &attr );
 
         XChangeWindowAttributes( data->display, data->whole_window, mask, &attr );
+#ifdef HAVE_X11_EXTENSIONS_XINPUT2_H
         x11drv_xinput2_enable( data->display, data->whole_window );
+#endif
     }
 }
 
@@ -1113,6 +1115,10 @@ static void set_initial_wm_hints( Display *display, Window window )
     /* class hints */
     if ((class_hints = XAllocClassHint()))
     {
+#ifdef __ANDROID__
+        class_hints->res_name = process_name;
+        class_hints->res_class = process_name;
+#else
         static char steam_proton[] = "steam_proton";
         const char *app_id = getenv("SteamAppId");
         char proton_app_class[128];
@@ -1125,17 +1131,20 @@ static void set_initial_wm_hints( Display *display, Window window )
             class_hints->res_name = steam_proton;
             class_hints->res_class = steam_proton;
         }
-
+#endif
         XSetClassHint( display, window, class_hints );
         XFree( class_hints );
     }
 
     /* set the WM_CLIENT_MACHINE and WM_LOCALE_NAME properties */
     XSetWMProperties(display, window, NULL, NULL, NULL, 0, NULL, NULL, NULL);
+
+#ifndef __ANDROID__
     /* set the pid. together, these properties are needed so the window manager can kill us if we freeze */
     i = getpid();
     XChangeProperty(display, window, x11drv_atom(_NET_WM_PID),
                     XA_CARDINAL, 32, PropModeReplace, (unsigned char *)&i, 1);
+#endif
 
     XChangeProperty( display, window, x11drv_atom(XdndAware),
                      XA_ATOM, 32, PropModeReplace, (unsigned char*)&dndVersion, 1 );
@@ -2085,6 +2094,18 @@ void set_net_active_window( HWND hwnd, HWND previous )
     XSendEvent( data->display, DefaultRootWindow( data->display ), False,
                 SubstructureRedirectMask | SubstructureNotifyMask, &xev );
     XFlush( data->display );
+
+#ifdef __ANDROID__
+    DWORD pid = 0;
+
+    NtUserGetWindowThread( hwnd, &pid );
+
+    XChangeProperty(data->display, window, x11drv_atom(_NET_WM_PID),
+                    XA_CARDINAL, 32, PropModeReplace, (unsigned char *)&pid, 1);
+
+    XChangeProperty( data->display, window, x11drv_atom(_NET_WM_HWND),
+                      XA_CARDINAL, 32, PropModeReplace, (unsigned char *)&hwnd, 2 );
+#endif
 }
 
 BOOL window_has_pending_wm_state( HWND hwnd, UINT state )
@@ -2507,7 +2528,9 @@ static void create_whole_window( struct x11drv_win_data *data )
 
     /* Set override-redirect attribute only after window creation, Mutter gets confused otherwise */
     window_set_managed( data, is_window_managed( data->hwnd, SWP_NOACTIVATE, FALSE ), FALSE );
+#ifdef HAVE_X11_EXTENSIONS_XINPUT2_H
     x11drv_xinput2_enable( data->display, data->whole_window );
+#endif
     set_initial_wm_hints( data->display, data->whole_window );
     set_wm_hints( data, 0 );
 
@@ -2827,8 +2850,10 @@ BOOL X11DRV_CreateWindow( HWND hwnd )
             if (!wcscmp( winstation_name, winsta0 ))
             {
                 /* listen to raw xinput event in the desktop window thread */
+#ifdef HAVE_X11_EXTENSIONS_XINPUT2_H
                 data->xinput2_rawinput = TRUE;
                 x11drv_xinput2_enable( data->display, DefaultRootWindow( data->display ) );
+#endif
             }
         }
         /* create the cursor clipping window */
