diff --git a/dlls/winex11.drv/opengl.c b/dlls/winex11.drv/opengl.c
index 478e3c2..b0619f8 100644
--- a/dlls/winex11.drv/opengl.c
+++ b/dlls/winex11.drv/opengl.c
@@ -664,6 +664,10 @@ static void *opengl_handle;
 static void init_opengl(void)
 {
     int error_base, event_base;
+#ifdef __ANDROID__
+    const char *force_glx_env;
+    int wine_x11forceglx = 0;
+#endif
     unsigned int i;
 
     /* No need to load any other libraries as according to the ABI, libGL should be self-sufficient
@@ -811,7 +815,14 @@ static void init_opengl(void)
 
     if(!X11DRV_WineGL_InitOpenglInfo()) goto failed;
 
+#ifdef __ANDROID__
+    force_glx_env = getenv("WINE_X11FORCEGLX");
+    if (force_glx_env)
+        wine_x11forceglx = atoi(force_glx_env);
+
+    if (XQueryExtension( gdi_display, "GLX", &glx_opcode, &event_base, &error_base ) || wine_x11forceglx)
+#else
     if (XQueryExtension( gdi_display, "GLX", &glx_opcode, &event_base, &error_base ))
+#endif
     {
         TRACE("GLX is up and running error_base = %d\n", error_base);
     } else {
